### PR TITLE
Use QR codes for other device authentication

### PR DESCRIPTION
**Pull request by span**
This commit adds support for QR codes. By default it is only used
for 'other device' authentication in line with the BankID
relying party guidelines.

It uses the auto login functionality since using a QR code follows
the same initialization steps as the auto launch. No PIN is
required and no PIN change is allowed.

A new dependency is added to create the QR codes. It is
a nuget package named 'QRCoder'[1].

Ref ActiveLogin/ActiveLogin.Authentication#19

[1] https://github.com/codebude/QRCoder/

### CODE REVIEW

File: src/ActiveLogin.Authentication.BankId.AspNetCore/Areas/BankIdAuthentication/Models/BankIdLoginApiInitializeResponse.cs [hunk 1]
```diff
@@ -2,13 +2,20 @@
 {
     public class BankIdLoginApiInitializeResponse
     {
-        internal BankIdLoginApiInitializeResponse(bool isAutoLaunch, bool showLaunchButton, bool checkStatus, string orderRef, string redirectUri)
+        internal BankIdLoginApiInitializeResponse(
```
Comment (PeterOrneholm) [inline src/ActiveLogin.Authentication.BankId.AspNetCore/Areas/BankIdAuthentication/Models/BankIdLoginApiInitializeResponse.cs]: As this is internal, we have full control, so I would skip the default`null` value.

File: src/ActiveLogin.Authentication.BankId.AspNetCore/Areas/BankIdAuthentication/Views/BankId/_LoginScript.cshtml [hunk 2]
```diff
@@ -23,6 +23,7 @@
         // Elements
 
         var startBankIdAppButtonElement = document.querySelector('#bankIdLoginStartApp');
+        var qrCodeElement = document.querySelector("#qr-code");
```
Comment (PeterOrneholm) [inline src/ActiveLogin.Authentication.BankId.AspNetCore/Areas/BankIdAuthentication/Views/BankId/_LoginScript.cshtml]: I would like this not to be an ID, as ID:s can collide. Please give that image element a class of something like 'qr-code-image` and find it by `statusElement.querySelector(´qr-code-image´);`

File: src/ActiveLogin.Authentication.BankId.AspNetCore/Areas/BankIdAuthentication/Views/BankId/_LoginStatus.cshtml [hunk 3]
```diff
@@ -3,9 +3,12 @@
 @inject IStringLocalizer<ActiveLogin.Authentication.BankId.AspNetCore.BankIdAuthenticationHandler> Localizer
 
 <div class="card" id="bankIdLoginStatus" style="display: @(Model.AutoLogin ? "block" : "none");">
+
+    <img src="#" title="qr-code" alt="qr-code" id="qr-code" style="display: none;" />
```
Comment (PeterOrneholm) [inline src/ActiveLogin.Authentication.BankId.AspNetCore/Areas/BankIdAuthentication/Views/BankId/_LoginStatus.cshtml]: Remove the title tag, and set alt tag to a localize text. We dont need title, but should use alt for accessibility reasdons.

Swedish: QR-kod för BankID
English: QR Code for BankID

File: src/ActiveLogin.Authentication.BankId.AspNetCore/Models/BankIdLoginOptions.cs [hunk 4]
```diff
@@ -5,13 +5,20 @@ namespace ActiveLogin.Authentication.BankId.AspNetCore.Models
...
     public class BankIdLoginOptions
     {
-        public BankIdLoginOptions(List<string> certificatePolicies, SwedishPersonalIdentityNumber personalIdentityNumber, bool allowChangingPersonalIdentityNumber, bool autoLaunch, bool allowBiometric)
+        public BankIdLoginOptions(
+            List<string> certificatePolicies,
+            SwedishPersonalIdentityNumber personalIdentityNumber,
+            bool allowChangingPersonalIdentityNumber,
+            bool autoLaunch,
+            bool allowBiometric,
+            bool idUseQrCode)
```
Comment (PeterOrneholm) [inline src/ActiveLogin.Authentication.BankId.AspNetCore/Models/BankIdLoginOptions.cs]: I think "Bank" fell of here.

File: src/ActiveLogin.Authentication.BankId.AspNetCore/Models/BankIdQrCodeGenerator.cs [hunk 5]
```diff
@@ -0,0 +1,19 @@
+using QRCoder;
+
+namespace ActiveLogin.Authentication.BankId.AspNetCore.Models
+{
+    public class BankIdQrCodeGenerator : IBankIdQrCodeGenerator
```
Comment (PeterOrneholm) [inline src/ActiveLogin.Authentication.BankId.AspNetCore/Models/BankIdQrCodeGenerator.cs]: I would call this `QRCoderBankIdQrCodeGenerator` so it's clear what implementation it uses.
Also, please write a small "intellisense documentation" with one line of what it does.

File: src/ActiveLogin.Authentication.BankId.AspNetCore/Models/BankIdQrCodeGenerator.cs [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+        public string GenerateQrCodeAsBase64(string autoStartToken)
+        {
+            QRCodeGenerator qrGenerator = new QRCodeGenerator();
+
+            var qrUrl = $"bankid:///?autostarttoken={autoStartToken}";
+            QRCodeData qrCodeData = qrGenerator.CreateQrCode(qrUrl, QRCodeGenerator.ECCLevel.Q);
+
+            Base64QRCode qrCode = new Base64QRCode(qrCodeData);
+
+            return qrCode.GetGraphic(20);
```
Comment (PeterOrneholm) [inline src/ActiveLogin.Authentication.BankId.AspNetCore/Models/BankIdQrCodeGenerator.cs]: Make the number 20 a constant, so it's easy to see what it represents. `pixelsPerModule`, right?

File: src/ActiveLogin.Authentication.BankId.AspNetCore/Models/BankIdQrCodeGenerator.cs [hunk 7]
```diff
@@ -0,0 +1,19 @@
...
+
+namespace ActiveLogin.Authentication.BankId.AspNetCore.Models
+{
+    public class BankIdQrCodeGenerator : IBankIdQrCodeGenerator
+    {
+        public string GenerateQrCodeAsBase64(string autoStartToken)
+        {
+            QRCodeGenerator qrGenerator = new QRCodeGenerator();
+
+            var qrUrl = $"bankid:///?autostarttoken={autoStartToken}";
```
Comment (PeterOrneholm) [inline src/ActiveLogin.Authentication.BankId.AspNetCore/Models/BankIdQrCodeGenerator.cs]: Please use the `QueryBuilder` as we used to compose theurl in `BankIdLauncher`. This ensures we do encoding correct etc.  But as you have done, the only param we have is `autostarttoken`. So just the query building part.

File: src/ActiveLogin.Authentication.BankId.AspNetCore/Models/IBankIdQrCodeGenerator.cs [hunk 8]
```diff
@@ -0,0 +1,7 @@
+namespace ActiveLogin.Authentication.BankId.AspNetCore.Models
+{
+    public interface IBankIdQrCodeGenerator
```
Comment (PeterOrneholm) [inline src/ActiveLogin.Authentication.BankId.AspNetCore/Models/IBankIdQrCodeGenerator.cs]: Please write a small "intellisense documentation" with one line of what it does.

File: src/ActiveLogin.Authentication.BankId.AspNetCore/Serialization/BankIdLoginOptionsSerializer.cs [hunk 9]
```diff
@@ -26,6 +26,7 @@ public byte[] Serialize(BankIdLoginOptions model)
                     writer.Write(model.AllowChangingPersonalIdentityNumber);
```
Comment (PeterOrneholm) [inline src/ActiveLogin.Authentication.BankId.AspNetCore/Serialization/BankIdLoginOptionsSerializer.cs]: At the top of this file, there is a constant that needs to be changed whenever we change the content of our opions.
Make it a 2 :)

`private const int FormatVersion = 2;`

File: src/ActiveLogin.Authentication.BankId.AspNetCore/Areas/BankIdAuthentication/Models/BankIdLoginApiInitializeResponse.cs (hunk 1 above)
Comment (span) [reply src/ActiveLogin.Authentication.BankId.AspNetCore/Areas/BankIdAuthentication/Models/BankIdLoginApiInitializeResponse.cs]: Sure! It was added as a way of skipping passing null in the builder methods.

File: src/ActiveLogin.Authentication.BankId.AspNetCore/Areas/BankIdAuthentication/Views/BankId/_LoginScript.cshtml (hunk 2 above)
Comment (span) [reply src/ActiveLogin.Authentication.BankId.AspNetCore/Areas/BankIdAuthentication/Views/BankId/_LoginScript.cshtml]: oki doki!

File: src/ActiveLogin.Authentication.BankId.AspNetCore/Areas/BankIdAuthentication/Views/BankId/_LoginStatus.cshtml (hunk 3 above)
Comment (span) [reply src/ActiveLogin.Authentication.BankId.AspNetCore/Areas/BankIdAuthentication/Views/BankId/_LoginStatus.cshtml]: of course!

File: src/ActiveLogin.Authentication.BankId.AspNetCore/Models/BankIdLoginOptions.cs (hunk 4 above)
Comment (span) [reply src/ActiveLogin.Authentication.BankId.AspNetCore/Models/BankIdLoginOptions.cs]: It did. Looking at it with fresh eyes I think we remove the `id` part as well. The context `BankId` is given by the class.

File: src/ActiveLogin.Authentication.BankId.AspNetCore/Models/BankIdQrCodeGenerator.cs (hunk 5 above)
Comment (span) [reply src/ActiveLogin.Authentication.BankId.AspNetCore/Models/BankIdQrCodeGenerator.cs]: I agree completely with the naming. Adding some documentation as well.

File: src/ActiveLogin.Authentication.BankId.AspNetCore/Models/BankIdQrCodeGenerator.cs (hunk 6 above)
Comment (span) [reply src/ActiveLogin.Authentication.BankId.AspNetCore/Models/BankIdQrCodeGenerator.cs]: Yes. No magic numbers please :)

File: src/ActiveLogin.Authentication.BankId.AspNetCore/Models/BankIdQrCodeGenerator.cs (hunk 7 above)
Comment (span) [reply src/ActiveLogin.Authentication.BankId.AspNetCore/Models/BankIdQrCodeGenerator.cs]: Will do!

File: src/ActiveLogin.Authentication.BankId.AspNetCore/Models/IBankIdQrCodeGenerator.cs (hunk 8 above)
Comment (span) [reply src/ActiveLogin.Authentication.BankId.AspNetCore/Models/IBankIdQrCodeGenerator.cs]: ok!

File: src/ActiveLogin.Authentication.BankId.AspNetCore/Serialization/BankIdLoginOptionsSerializer.cs (hunk 9 above)
Comment (span) [reply src/ActiveLogin.Authentication.BankId.AspNetCore/Serialization/BankIdLoginOptionsSerializer.cs]: :O

Comment (PeterOrneholm) [reply src/ActiveLogin.Authentication.BankId.AspNetCore/Serialization/BankIdLoginOptionsSerializer.cs]: Hehe, it's the convention Microsoft uses. The first field serialized is always a version. If someone has a browser open during release, and we make a breaking change to the "schema", the version ensure we can fail that request.

Comment (span) [reply src/ActiveLogin.Authentication.BankId.AspNetCore/Serialization/BankIdLoginOptionsSerializer.cs]: Very nice! I come from java land and I don't think we have conventions like that there 🤔 

.net core-land is much nicer :D

**Comment by span** [comment]
Setting this to ready to review now after a couple of great code reviews today. Very nice!

**Comment by PeterOrneholm** [comment]
Pushed a few minor things myself :)
Lets get this merged into master (after the cancel PR), and then I could separate the QR dependency into its own package later on.

**Comment by span** [comment]
Working on the conflicts.

File: src/ActiveLogin.Authentication.BankId.AspNetCore/Areas/BankIdAuthentication/Views/BankId/_LoginStatus.cshtml [hunk 10]
```diff
@@ -10,6 +10,8 @@
         <div class="status-spinner">
             <partial name="_spinner"/>
         </div>
+
+        <div class="status-message">@Model.LoginScriptOptions.InitialStatusMessage</div>
```
Comment (span) [inline src/ActiveLogin.Authentication.BankId.AspNetCore/Areas/BankIdAuthentication/Views/BankId/_LoginStatus.cshtml]: Not sure where this came from 🤔

Comment (span) [reply src/ActiveLogin.Authentication.BankId.AspNetCore/Areas/BankIdAuthentication/Views/BankId/_LoginStatus.cshtml]: Was there all along it seems :P

Comment (PeterOrneholm) [reply src/ActiveLogin.Authentication.BankId.AspNetCore/Areas/BankIdAuthentication/Views/BankId/_LoginStatus.cshtml]: :)

**Comment by span** [comment]
This should be all done by now. It will probably cause conflicts with other pr's as they are in the same classes poking around. We'll solve it as we go :)

File: src/ActiveLogin.Authentication.BankId.AspNetCore/Areas/BankIdAuthentication/Views/BankId/_LoginStyle.cshtml [hunk 11]
```diff
@@ -10,6 +10,11 @@
         transform: translate(-50%, -50%);
     }
 
+    .qr-code-image {
```
Comment (PeterOrneholm) [inline src/ActiveLogin.Authentication.BankId.AspNetCore/Areas/BankIdAuthentication/Views/BankId/_LoginStyle.cshtml]: Remove custom css, see explanation above.

File: test/ActiveLogin.Authentication.BankId.AspNetCore.Test/BankId_Authentication_Tests.cs [hunk 12]
```diff
@@ -164,16 +164,18 @@ public async Task BankIdAuthentication_Login_Returns_Form_And_Status()
 
             // Assert
             Assert.Equal(HttpStatusCode.OK, transaction.StatusCode);
+
             var content = await transaction.Content.ReadAsStringAsync();
             Assert.Contains("<form id=\"bankIdLoginForm\">", content);
-            Assert.Contains("<div id=\"bankIdLoginStatus\"", content);
+            Assert.Contains("<div class=\"card\" id=\"bankIdLoginStatus\" style=\"display: block;\">", content);
```
Comment (PeterOrneholm) [inline test/ActiveLogin.Authentication.BankId.AspNetCore.Test/BankId_Authentication_Tests.cs]: Don't need to change this now, but I recently stumbled upon [AngleSharp ](https://anglesharp.github.io/) for parsing HTML in C#. Microsoft seems to recommend it here: https://docs.microsoft.com/en-us/aspnet/core/test/integration-tests?view=aspnetcore-3.0

If I get it right, you can parse the DOM using the lib, query for the element you like and check a specific prop. That would make our tests less fragile.

File: src/ActiveLogin.Authentication.BankId.AspNetCore/Areas/BankIdAuthentication/Views/BankId/_LoginStatus.cshtml [hunk 13]
```diff
@@ -2,15 +2,16 @@
 @model ActiveLogin.Authentication.BankId.AspNetCore.Areas.BankIdAuthentication.Models.BankIdLoginViewModel
 @inject IStringLocalizer<ActiveLogin.Authentication.BankId.AspNetCore.BankIdAuthenticationHandler> Localizer
 
+<div class="card" id="bankIdLoginStatus" style="display: @(Model.AutoLogin ? "block" : "none");">
```
[the remark this answers is not in the document]
Comment (span) [reply src/ActiveLogin.Authentication.BankId.AspNetCore/Areas/BankIdAuthentication/Views/BankId/_LoginStatus.cshtml]: Very ncie. Hope I get it right this time :P

File: test/ActiveLogin.Authentication.BankId.AspNetCore.Test/BankId_Authentication_Tests.cs (hunk 12 above)
Comment (span) [reply test/ActiveLogin.Authentication.BankId.AspNetCore.Test/BankId_Authentication_Tests.cs]: Very nice! I created an issue for it #147

**Comment by span** [comment]
Pushed a new commit with requested changes.

When testing a new issue popped out at me. When the different messages are displayed, the qr code jumps around since the layout resizes from the vertical center and not from the vertical top.

Will this be confusing for users?

File: src/ActiveLogin.Authentication.BankId.AspNetCore/Areas/BankIdAuthentication/Views/BankId/_LoginScript.cshtml [hunk 14]
```diff
@@ -4,7 +4,6 @@
     (function (options) {
         // Pre check
         var statusElement = document.querySelector('#bankIdLoginStatus');
-        var statusWrapperElement = statusElement.querySelector('.status-wrapper');
```
Comment (PeterOrneholm) [inline src/ActiveLogin.Authentication.BankId.AspNetCore/Areas/BankIdAuthentication/Views/BankId/_LoginScript.cshtml]: Just realized we need to add back the `.status-wrapper` class at the `.card` element, and modify the JS back to apply styling etc on that element.

File: src/ActiveLogin.Authentication.BankId.AspNetCore/Areas/BankIdAuthentication/Views/BankId/_LoginScript.cshtml [hunk 15]
```diff
@@ -213,7 +222,7 @@
                 textClass = '';
             }
 
-            statusWrapperElement.className = 'card status-wrapper bg-' + type + ' ' + textClass;
```
Comment (PeterOrneholm) [inline src/ActiveLogin.Authentication.BankId.AspNetCore/Areas/BankIdAuthentication/Views/BankId/_LoginScript.cshtml]: Should be `statusWrapperElement` again, se above.

**Comment by span** [comment]
Fixed the status wrapper. I also realised the margin between the status card and the qr code had been lost somewhere so I re-added it. I think it looks cleaner with a small separation between the two.

With margin:
Screenshot 2019-11-03 at 08 16 11


Without margin:
Screenshot 2019-11-03 at 08 13 20

**Comment by PeterOrneholm** [comment]
> I think it looks cleaner with a small separation between the two.

Totally agree :)

**Comment by PeterOrneholm** [comment]
Nice work!

One minor thing to change, and one question.

I think we should change the class `'light'` into `'`white'`, that would align better with the white in the QR code. I found four occurrences. can you fix this?

Light:
light

White:
white


The question. Do you know why BankID fails so fast if you don't open your app fast enough? After 25 seconds I get this:

> The BankID app couldn’t be found on your computer or mobile device. Please install it and order a BankID from your internet bank. Install the app from your app store or https://install.bankid.com.

25 seconds isn't that much. If that is a limit from BankID, I think we should handle that by having the Cancel button initiate a new code. But let's get this merged first, and then have a look at the cancel flow after error in the other PR.

**Comment by PeterOrneholm** [comment]
Ahh, just found this.

image

We do display RFA17 (A), but if QR code is enabled, we should display RFA17 (B). Can you have a look at how to show this error instead?

**Comment by span** [comment]
Added the light theme instead and also a first shot at the other message. I have not had the time to test yet unfortunately so it most likely does not work as is but it will perhaps give you an idea if you like the approach or not.

File: test/ActiveLogin.Authentication.BankId.Api.Test/UserMessage/BankIdRecommendedUserMessage_CollectResponse_Tests.cs [hunk 16]
```diff
@@ -19,13 +19,14 @@ public BankIdRecommendedUserMessage_CollectResponse_Tests()
         [InlineData(CollectHintCode.Started, MessageShortName.RFA15A)]
         [InlineData(CollectHintCode.UserSign, MessageShortName.RFA9)]
         [InlineData(CollectHintCode.Unknown, MessageShortName.RFA21)]
-        public void GetMessageShortNameForCollectResponse_ShouldReturnReccomendedMessageForCollectResponse_PendingOrder_BankID_AutomaticallyStartBankIDAPP(CollectHintCode collectHintCode, MessageShortName expected)
+        public void GetMessageShortNameForCollectResponse_ShouldReturnRecommendedMessageForCollectResponse_PendingOrder_BankID_AutomaticallyStartBankIDAPP(CollectHintCode collectHintCode, MessageShortName expected)
```
Comment (span) [inline test/ActiveLogin.Authentication.BankId.Api.Test/UserMessage/BankIdRecommendedUserMessage_CollectResponse_Tests.cs]: Sorry but I had to rename these :/

**Comment by smuda** [comment]
When requireAutostartToken is used the timeout for BankID is currently 30 seconds. Checkout https://demo.bankid.com/Logon.aspx?environment=Produktion where they handle the timeout by generating a new transaction automatically up to five times.

File: test/ActiveLogin.Authentication.BankId.Api.Test/UserMessage/BankIdRecommendedUserMessage_CollectResponse_Tests.cs (hunk 16 above)
Comment (PeterOrneholm) [reply test/ActiveLogin.Authentication.BankId.Api.Test/UserMessage/BankIdRecommendedUserMessage_CollectResponse_Tests.cs]: Looks good :)

**Comment by PeterOrneholm** [comment]
> When requireAutostartToken is used the timeout for BankID is currently 30 seconds. Checkout https://demo.bankid.com/Logon.aspx?environment=Produktion where they handle the timeout by generating a new transaction automatically up to five times.

Very good input @smuda! I'd say we should to the same. 

@span: I think the retry strategy mentioned above would be kind of coupled to our cancel. I'd say our strategy should be:

1. Fix the last things on this regrading styling, and then merge to master.
2. Finish the PR for fixing Cancel (#145). While on that, we will make sure that cancel in QR behaves as expected.
3. Make another PR implementing 5 retries for QR.

By doing it step by step we don't end up with gigantic PR never merged,

From my point of view, this PR looks good now. If you agree, I will merge this? @span

**Comment by span** [comment]
@PeterOrneholm Sure, go ahead! I still have not performed proper testing of the new message handling though.

**Comment by PeterOrneholm** [comment]
> @PeterOrneholm Sure, go ahead! I still have not performed proper testing of the new message handling though.

I'll give it a try before merging. Could you write a small note in the readme on how to use QR?
It's on by default, so the "customize" section here is a relevant place.
https://github.com/ActiveLogin/ActiveLogin.Authentication/blob/master/docs/getting-started-bankid.md#customizing-bankid

File: docs/getting-started-bankid.md [hunk 17]
```diff
@@ -211,11 +211,14 @@ BankId options allows you to set and override some options such as these.
     // Limit possible login methods to, for example, only allow BankID on smartcard.
     options.BankIdCertificatePolicies = BankIdCertificatePolicies.GetPoliciesForProductionEnvironment(...);
 
-	// Issue birthdate claim based on data extracted from the personal identity number
```
Comment (span) [inline docs/getting-started-bankid.md]: The following lines had white space issues where tabs were not presented in the same way as spaces in the markdown-render.

**Comment by PeterOrneholm** [comment]
Yaay, we are merged! Let's look at the Cancel PR now, and get that flow working.

File: docs/getting-started-bankid.md [hunk 18]
```diff
@@ -217,8 +217,9 @@ BankId options allows you to set and override some options such as these.
     // Issue gender claim based on data extracted from the personal identity number
     options.IssueGenderClaim = true;
 
-    // Turn off qr code and use personal identity number instead
-    options.BankIdUseQrCode = false
+    // Turn off 
+    code and use personal identity number instead
```
Comment (span) [inline docs/getting-started-bankid.md]: Looks strange, does this compile?

**Pull request merged**